sdks/java/core/src/main/java/org/apache/beam/sdk/io/DynamicFileDestinations.java

```diff
@@ -63,18 +80,142 @@ public void populateDisplayData(DisplayData.Builder builder) {
     }
   }
 
+  // TODO: These two classes are duplicated in AvroCoder. Is there a common place to put them?
+  private static class SerializableSchemaString implements Serializable {
+    private final String schema;
+    private SerializableSchemaString(String schema) {
+      this.schema = schema;
+    }
+
+    private Object readResolve() throws IOException, ClassNotFoundException {
+      return new SerializableSchemaSupplier(Schema.parse(schema));
+    }
+  }
+
+  private static class SerializableSchemaSupplier implements Serializable, Supplier<Schema> {
+    private final Schema schema;
+    private SerializableSchemaSupplier(Schema schema) {
+      this.schema = schema;
+    }
+
+    private Object writeReplace() {
+      return new SerializableSchemaString(schema.toString());
+    }
+
+    @Override
+    public Schema get() {
+      return schema;
+    }
+  }
+
+  /**
+   * Always returns a constant {@link FilenamePolicy}, {@link Schema}, metadata, and codec.
+   */
+  private static class ConstantAvroDestination<UserT, OutputT>
```

Comment:
it could be, but it feels weird since that is just an abstract base class for users to override. We could move things out, but IMO we should then create anoter class